pmc_heu.cpp

```diff
@@ -17,96 +17,116 @@
     }
     else if (sz > mc)
         mc = sz;
     return;
 }
 
-int pmc_heu::search_bounds(pmc_graph& G,
-        vector<int>& C_max) {
+int pmc_heu::search_bounds(const pmc_graph& G,
+        std::vector<int>& C_max) {
+
+    V = &G.get_vertices();
+    E = &G.get_edges();
+
+    std::vector<int> C;
+    std::vector<Vertex> P;
 
-    V = G.get_vertices();
-    E = G.get_edges();
-    degree = G.get_degree();
-    vector <int> C, X;
-    C.reserve(ub);
     C_max.reserve(ub);
-    vector<Vertex> P, T;
+    C.reserve(ub);
     P.reserve(G.get_max_degree()+1);
-    T.reserve(G.get_max_degree()+1);
-    vector<short> ind(G.num_vertices(),0);
+
+    bool_vector ind(G.num_vertices(), false);
 
     bool found_ub = false;
-    int mc = 0, mc_prev, mc_cur, i, v, k, lb_idx = 0;
+    int mc = 0;
 
     #pragma omp parallel for schedule(dynamic) \
-        shared(G, X, mc, C_max, lb_idx) private(i, v, P, mc_prev, mc_cur, C, k) firstprivate(ind) \
+        shared(G, mc, C_max, found_ub) \
+        private(P, C) firstprivate(ind) \
         num_threads(num_threads)
-    for (i = G.num_vertices()-1; i >= 0; --i) {
-        if (found_ub) continue;
+    for (int i = G.num_vertices()-1; i >= 0; --i) {
+        bool found_ub_local = false;
+        #pragma omp atomic read acquire
+        found_ub_local = found_ub;
 
-        v = (*order)[i];
+        if (found_ub_local) {
+            continue;
+        }
+
+        const int v = (*order)[i];
+
+        int mc_prev, mc_cur;
+        #pragma omp atomic read acquire
         mc_prev = mc_cur = mc;
 
-        if ((*K)[v] > mc) {
+        if ((*K)[v] > mc_cur) {
             for (long long j = (*V)[v]; j < (*V)[v + 1]; j++)
-                if ((*K)[(*E)[j]] > mc)
-                    P.push_back( Vertex((*E)[j], compute_heuristic((*E)[j])) );
-
+                if ((*K)[(*E)[j]] > mc_cur)
+                    P.emplace_back((*E)[j], compute_heuristic((*E)[j]));
 
             if (P.size() > mc_cur) {
                 std::sort(P.begin(), P.end(), incr_heur);
                 branch(P, 1 , mc_cur, C, ind);
 
+                if (mc_cur >= ub) {
+                    #pragma omp atomic write release
+                    found_ub = true;
+                }
+
                 if (mc_cur > mc_prev) {
-                    if (mc < mc_cur) {
+                    #pragma omp atomic read acquire
+                    mc_prev = mc;
+
+                    if (mc_cur > mc_prev) {
+                        #pragma omp atomic write release
+                        mc = mc_cur;
+
+                        C.push_back(v);
+
                         #pragma omp critical
-                        if (mc < mc_cur) {
-                            mc = mc_cur;
-                            C.push_back(v);
-                            C_max = C;
-                            if (mc >= ub) found_ub = true;
-                            print_info(C_max);
-                        }
+                        std::swap(C_max, C);
```

Comment:
I intended to avoid memory copy by swapping two reserved containers. `C` is cleared later but still reserved, so the next iteration can reuse it without reallocation.